sound/soc/soc-topology.c

```diff
@@ -2087,7 +2088,12 @@ static int soc_tplg_pcm_elems_load(struct soc_tplg *tplg,
 		}
 
 		/* create the FE DAIs and DAI links */
-		soc_tplg_pcm_create(tplg, _pcm);
+		ret = soc_tplg_pcm_create(tplg, _pcm);
+		if (ret < 0) {
+			if (!abi_match)
+				kfree(_pcm);
```

Comment:
@dragosht I'm not too sure about this one. The DAI_DRV that is created in soc_tplg_pcm_create could still have been added to the comp->dobj_list and when you attempt to unload the dai, it will point to a freed pcm. Is this correct?

Reply:
@ranj063 I see that topology _pcm copy is freed right after pcm_create() anyway, dai_drv doesn't seem to keep any references to it, so it shouldn't be used: https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/soc-topology.c#L2090 . I did the same thing here just to prevent it leak to away ...

Reply:
@dragosht dai_drc does not keep the reference but the SOF driver does in its spcm. So when dai_unload happens, we will try to access a freed pcm?

Reply:
@ranj063 Actually the sof spcm seems to keep its own copy of tplg_pcm, not a pointer to it: https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/sof-audio.h#L47 and it creates that in dai_load:  https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/topology.c#L2455

Reply:
@dragosht yes you are correct. Thanks for clarifying.